src/instructlab/llamacpp/convert_to_gguf.py

```diff
@@ -0,0 +1,1381 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Future
+from __future__ import annotations
+
+# Standard
+from enum import IntEnum
+from hashlib import sha256
+from itertools import chain
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    ContextManager,
+    Iterable,
+    Iterator,
+    Literal,
+    Sequence,
+    TypeVar,
+    cast,
+)
+import contextlib
+import json
+import logging
+import math
+import os
+import sys
+
+# Third Party
+import numpy as np
+import torch
+
+if TYPE_CHECKING:
+    # Third Party
+    from torch import Tensor
+
+if "NO_LOCAL_GGUF" not in os.environ:
+    sys.path.insert(1, str(Path(__file__).parent / "gguf-py"))
+# Third Party
+import gguf
+
+logger = logging.getLogger("hf-to-gguf")
+
+
+###### MODEL DEFINITIONS ######
+
+
+class SentencePieceTokenTypes(IntEnum):
+    NORMAL = 1
+    UNKNOWN = 2
+    CONTROL = 3
+    USER_DEFINED = 4
+    UNUSED = 5
+    BYTE = 6
+
+
+AnyModel = TypeVar("AnyModel", bound="type[Model]")
+
+
+class Model:
+    _model_classes: dict[str, type[Model]] = {}
+
+    dir_model: Path
+    ftype: gguf.LlamaFileType
+    fname_out: Path
+    is_big_endian: bool
+    endianess: gguf.GGUFEndian
+    use_temp_file: bool
+    lazy: bool
+    part_names: list[str]
+    is_safetensors: bool
+    hparams: dict[str, Any]
+    block_count: int
+    tensor_map: gguf.TensorNameMap
+    tensor_names: set[str] | None
+    gguf_writer: gguf.GGUFWriter
+    model_name: str | None
+    metadata_override: Path | None
+    dir_model_card: Path
+
+    # subclasses should define this!
+    model_arch: gguf.MODEL_ARCH
+
+    def __init__(
+        self,
+        dir_model: Path,
+        ftype: gguf.LlamaFileType,
+        fname_out: Path,
+        is_big_endian: bool = False,
+        use_temp_file: bool = False,
+        eager: bool = False,
+        metadata_override: Path | None = None,
+        model_name: str | None = None,
+        split_max_tensors: int = 0,
+        split_max_size: int = 0,
+        dry_run: bool = False,
+        small_first_shard: bool = False,
+        hparams: dict[str, Any] | None = None,
+    ):
+        if type(self) is Model:
+            raise TypeError(
+                f"{type(self).__name__!r} should not be directly instantiated"
+            )
+
+        self.dir_model = dir_model
+        self.ftype = ftype
+        self.fname_out = fname_out
```

Comment:
```suggestion
        self.fname_out = Path(fname_out)
```
To make code more robust and prevent similar error like described in https://github.com/instructlab/instructlab/pull/2950/files#r1975005755